// notion-rehype-k/dist/types.js
export var BlockType;
(function (BlockType) {
    BlockType["paragraph"] = "paragraph";
    BlockType["heading"] = "heading";
    BlockType["callout"] = "callout";
    BlockType["quote"] = "quote";
    BlockType["bulleted_list"] = "bulleted_list";
    BlockType["numbered_list"] = "numbered_list";
    BlockType["to_do_list"] = "to_do_list";
    // to_do = 'to_do', // uses to_do_list instead
    BlockType["toggle"] = "toggle";
    BlockType["code"] = "code";
    BlockType["embed"] = "embed";
    BlockType["image"] = "image";
    BlockType["video"] = "video";
    BlockType["file"] = "file";
    BlockType["bookmark"] = "bookmark";
    BlockType["equation"] = "equation";
    BlockType["divider"] = "divider";
    BlockType["column_list"] = "column_list";
    BlockType["column"] = "column";
    BlockType["link_preview"] = "link_preview";
    BlockType["link_to_page"] = "link_to_page";
    BlockType["synced_block"] = "synced_block";
    BlockType["table"] = "table";
    BlockType["audio"] = "audio";
    // table_row = 'table_row', // intergrated in table handler
    // TODO: These types need to be implemented
    // child_page = 'child_page',
    // child_database = 'child_database',
    // pdf = 'pdf',
    // table_of_contents = 'table_of_contents',
    // template = 'template', // May never be implemented
})(BlockType || (BlockType = {}));
