// @ascorbic/csv-loader/dist/index.js
// src/csv-loader.ts
import { createReadStream, existsSync } from "fs";
import { fileURLToPath } from "url";
import { relative } from "pathe";
import Papa from "papaparse";
import { AstroError } from "astro/errors";
var camelize = (str) => str.split(/[-_\s]+/).map(
  (word, index) => `${index === 0 ? word.charAt(0).toLowerCase() : word.charAt(0).toUpperCase()}${word.slice(1)}`
).join("");
function csvLoader({
  fileName,
  idField,
  transformHeader = camelize,
  parserOptions
}) {
  async function syncData(filePath, { logger, parseData, store, config }) {
    const relativePath = relative(fileURLToPath(config.root), filePath);
    const csvStream = Papa.parse(Papa.NODE_STREAM_INPUT, {
      dynamicTyping: true,
      ...parserOptions,
      header: true,
      transformHeader: transformHeader === false ? void 0 : transformHeader
    });
    createReadStream(filePath).pipe(csvStream);
    await new Promise((resolve, reject) => {
      let count = 0;
      let row = 0;
      store.clear();
      csvStream.on("end", () => {
        logger.info(`Loaded ${count} entries from ${fileName}`);
        resolve();
      });
      csvStream.on("error", (error) => {
        logger.error("Error reading data");
        reject(error);
      });
      csvStream.on("data", async (data) => {
        if (!idField) {
          idField = Object.keys(data)[0];
          logger.info(`No ID field specified, using first column: ${idField}`);
          if (!idField) {
            csvStream.end();
            reject("No ID field found in CSV file");
            return;
          }
        }
        row++;
        if (!(idField in data) || !(data[idField] && data[idField] !== 0)) {
          logger.warn(`No ID (${idField}) found in row ${row}. Skipping.`);
          return;
        }
        count++;
        const id = String(data[idField]);
        const parsedData = await parseData({
          id,
          data,
          filePath: relativePath
        });
        store.set({ id, data: parsedData, filePath: relativePath });
      });
    }).catch((error) => {
      logger.error("Error reading data: " + error);
      throw new AstroError("Aborted loading CSV data");
    });
  }
  return {
    name: "csv-loader",
    load: async (options) => {
      const { config, logger, watcher } = options;
      logger.info(`Loading CSV data from ${fileName}`);
      const url = new URL(fileName, config.root);
      if (!existsSync(url)) {
        logger.error(`File not found: ${fileName}`);
        return;
      }
      const filePath = fileURLToPath(url);
      await syncData(filePath, options);
      watcher?.on("change", async (changedPath) => {
        if (changedPath === filePath) {
          logger.info(`Reloading data from ${fileName}`);
          await syncData(filePath, options);
        }
      });
    }
  };
}
export {
  csvLoader
};
